### PR TITLE
types: fix type checking errors in dvc.experiments

### PR DESCRIPTION
**Pull request by skshetry**
Part of #4697 
* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

File: dvc/repo/experiments/__init__.py [hunk 1]
```diff
@@ -375,11 +375,9 @@ def _commit(
         if not self.scm.is_dirty(untracked_files=True):
             raise UnchangedExperimentError(self.scm.get_rev())
 
-        rev = self.scm.get_rev()
-        if not baseline_rev:
-            baseline_rev = rev
-        checkpoint = "-checkpoint" if checkpoint else ""
-        exp_name = f"{baseline_rev[:7]}-{exp_hash}{checkpoint}"
+        rev = baseline_rev or self.scm.get_rev()
+        suffix = "-checkpoint" if checkpoint else ""
```
Comment (skshetry) [inline dvc/repo/experiments/__init__.py:379]: `checkpoint` is already defined as being `bool`. So, changed the variable name here.

File: dvc/repo/experiments/__init__.py [hunk 2]
```diff
@@ -450,7 +448,7 @@ def new(
         """
         if checkpoint_resume is not None:
             return self._resume_checkpoint(
-                *args, **kwargs, checkpoint_resume=checkpoint_resume
+                *args, checkpoint_resume=checkpoint_resume, **kwargs
```
Comment (skshetry) [inline dvc/repo/experiments/__init__.py:451]: Mypy was complaining that `checkpoint_resume` might be set twice. Not sure how, but changed the position and it stopped.

File: dvc/repo/experiments/__init__.py [hunk 3]
```diff
@@ -628,7 +626,7 @@ def _reproduce(self, executors: dict, jobs: Optional[int] = 1) -> dict:
 
         Returns dict containing successfully executed experiments.
         """
-        result = {}
+        result: dict = {}
```
Comment (skshetry) [inline dvc/repo/experiments/__init__.py:629]: I was not sure specifically what type it is. So, just left a `dict` there. Please make it more specific later.

File: dvc/repo/experiments/__init__.py [hunk 4]
```diff
@@ -701,6 +699,7 @@ def _collect_executor(self, rev, executor, stages, result, lock):
         finally:
             lock.release()
 
+    @no_type_check
```
Comment (skshetry) [inline dvc/repo/experiments/__init__.py:702]: I was not sure of specific types in the below args/kwargs type. They have `Iterable` type but not sure of what. For now, disable type-checking in that function.

File: dvc/repo/experiments/executor.py [hunk 5]
```diff
@@ -41,6 +41,7 @@ def __init__(
         self.branch = branch
         self.repro_args = kwargs.pop("repro_args", [])
         self.repro_kwargs = kwargs.pop("repro_kwargs", {})
+        self.tmp_dir: Optional[TemporaryDirectory] = None
```
Comment (skshetry) [inline dvc/repo/experiments/executor.py:44]: Leaving up to @pmrowla  to decide what's better. I just did this to silence mypy in `logger.debug` calls made in the `Experiment`.

**Review by pmrowla — APPROVED** [review]
Thanks for fixing these @skshetry

**Pull request merged**